include/logging/log_core.h

```diff
@@ -162,9 +163,11 @@ extern "C" {
 
 #define Z_LOG_INTERNAL_X(N, ...)  UTIL_CAT(_LOG_INTERNAL_, N)(__VA_ARGS__)
 
-#define __LOG_INTERNAL(_src_level, ...)					 \
+#define __LOG_INTERNAL(is_user_context, _src_level, ...)		 \
 	do {								 \
-		if (IS_ENABLED(CONFIG_LOG_IMMEDIATE)) {		 \
+		if (is_user_context) {					 \
```

Comment:
why is this not enclosed in a `#if CONFIG_USERSPACE`? won't having the `if` statement add code even when userspace is disabled?

Reply:
This is a macro definition, where I cannot use #ifdefs. So if I would like to use this classic method, I would have to several definitions of the macros here.

As you noted, I a relying on the fact, that value returned by the `_is_user_context()` is known at compile time when userspace is disabled:
```
/**
 * Indicate whether the CPU is currently in user mode
 *
 * @return true if the CPU is currently running with user permissions
 */
static inline bool _is_user_context(void)
{
#ifdef CONFIG_USERSPACE
       return z_arch_is_user_context();
#else
       return false;
#endif
}
```
Unfortunatelly, the dead code is not removed when you are compiling with `CONFIG_NO_OPTIMIZATIONS=y`, but I assume that additional overhead will be not a problem in such case.